File: samples/java/ShoppingMall/Server/OrderWorkflow/src/main/java/systems/zlink/samples/shoppingmall/server/orderworkflow/spots/OrderWorkflowSpot.java
package systems.zlink.samples.shoppingmall.server.orderworkflow.spots;

import systems.zlink.framework.spots.ZLinkInstanceSpot;
import systems.zlink.framework.spots.ZLinkInstanceSpotContext;
import systems.zlink.samples.shoppingmall.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class OrderWorkflowSpot implements ZLinkInstanceSpot {
    private final ZLinkInstanceSpotContext context;

    public OrderWorkflowSpot(ZLinkInstanceSpotContext context) {
        this.context = context;
    }

    @Override
    public ZLinkInstanceSpotContext context() {
        return context;
    }

    // --8<-- [start:doc-sm-close-terminal]
    public CompletionStage<Void> closeIfTerminal(Messages.OrderState state) {
        if (isTerminal(state)) {
            return context.close().thenApply(closed -> null);
        }
        return CompletableFuture.completedFuture(null);
    }

    // --8<-- [end:doc-sm-close-terminal]

    public boolean isTerminal(Messages.OrderState state) {
        return Messages.OrderStatuses.Confirmed.equals(state.status())
                || Messages.OrderStatuses.Failed.equals(state.status());
    }
}

File: samples/kotlin/ShoppingMall/Server/OrderWorkflow/src/main/kotlin/systems/zlink/samples/kotlin/shoppingmall/server/orderworkflow/OrderWorkflowSpot.kt
package systems.zlink.samples.kotlin.shoppingmall.server.orderworkflow

import java.util.concurrent.CompletionStage
import systems.zlink.framework.spots.ZLinkInstanceSpot
import systems.zlink.framework.spots.ZLinkInstanceSpotContext
import systems.zlink.samples.kotlin.shoppingmall.shared.contracts.OrderState
import systems.zlink.samples.kotlin.shoppingmall.shared.contracts.OrderStatuses

class OrderWorkflowSpot(private val instanceContext: ZLinkInstanceSpotContext) : ZLinkInstanceSpot {
    override fun context(): ZLinkInstanceSpotContext = instanceContext

    fun requireOrder(orderId: String) {
        require(orderId == instanceContext.spotId()) {
            "request order does not match workflow Spot: $orderId"
        }
    }

    // --8<-- [start:doc-sm-close-terminal]
    fun closeIfTerminal(state: OrderState): CompletionStage<Void> =
        if (isTerminal(state)) {
            instanceContext.close().thenApply<Void> { null }
        } else {
            java.util.concurrent.CompletableFuture.completedFuture<Void>(null)
        }

    // --8<-- [end:doc-sm-close-terminal]

    fun isTerminal(state: OrderState): Boolean =
        state.status == OrderStatuses.Confirmed || state.status == OrderStatuses.Failed
}
